### PR TITLE
introduce caveat context in PermissionService

### PR DESCRIPTION
**Pull request by vroldanbet**
Part of https://github.com/authzed/spicedb/issues/881
Part of https://github.com/authzed/spicedb/issues/882
Part of https://github.com/authzed/spicedb/issues/880

this makes it possible for clients to provide context for caveat evaluation at request time.
This context will be injected at evaluation time and influence the evaluation result.

### CODE REVIEW

File: authzed/api/v1/permission_service.proto [hunk 1]
```diff
@@ -250,6 +251,9 @@ message CheckPermissionRequest {
 
   // subject is the subject that will be checked for the permission or relation.
   SubjectReference subject = 4 [ (validate.rules).message.required = true ];
+
+  /** context consists of named values that are injected into the caveat evaluation context **/
+  google.protobuf.Struct context = 5 [ (validate.rules).message.required = false ];
 }
 
 message CheckPermissionResponse {
```
Comment (josephschorr) [inline authzed/api/v1/permission_service.proto:259]: Please add `PERMISSIONSHIP_MAYBE_PERMISSION` and a `repeated string missing_required_context = 3;`

File: authzed/api/v1/permission_service.proto [hunk 2]
```diff
@@ -324,6 +328,9 @@ message LookupResourcesRequest {
 
   // subject is the subject with access to the resources.
   SubjectReference subject = 4 [ (validate.rules).message.required = true ];
+
+  /** context consists of named values that are injected into the caveat evaluation context **/
+  google.protobuf.Struct context = 5 [ (validate.rules).message.required = false ];
 }
 
 // LookupResourcesResponse contains a single matching resource object ID for the
```
Comment (josephschorr) [inline authzed/api/v1/permission_service.proto:336]: We need to add an enum here to indicate that the response is either `HAS_PERMISSION` or `MAYBE_PERMISSION`, along with the same "missing fields"

File: authzed/api/v1/permission_service.proto [hunk 3]
```diff
@@ -362,6 +369,9 @@ message LookupSubjectsRequest {
     pattern : "^([a-z][a-z0-9_]{1,62}[a-z0-9])?$",
     max_bytes : 64,
   } ];
+
+  /** context consists of named values that are injected into the caveat evaluation context **/
+  google.protobuf.Struct context = 6 [ (validate.rules).message.required = false ];
 }
 
 // LookupSubjectsResponse contains a single matching subject object ID for the
```
Comment (josephschorr) [inline authzed/api/v1/permission_service.proto:377]: We need to add an enum here to indicate that the response is either `HAS_PERMISSION` or `MAYBE_PERMISSION`, along with the same "missing fields"

File: authzed/api/v1/permission_service.proto [hunk 4]
```diff
@@ -250,13 +251,25 @@ message CheckPermissionRequest {
 
   // subject is the subject that will be checked for the permission or relation.
   SubjectReference subject = 4 [ (validate.rules).message.required = true ];
+
+  /** context consists of named values that are injected into the caveat evaluation context **/
+  google.protobuf.Struct context = 5 [ (validate.rules).message.required = false ];
+}
+
+// PartialCaveatInfo carries information necessary or the client to take action in the even a response was
```
Comment (josephschorr) [inline authzed/api/v1/permission_service.proto:259]: for the client

Comment (josephschorr) [inline authzed/api/v1/permission_service.proto:259]: in the event a response contains a partially evaluated caveat

File: .github/workflows/lint.yaml [hunk 5]
```diff
@@ -4,10 +4,9 @@ on:
   push:
     branches:
       - "!dependabot/*"
-      - "*"
+      - "main"
```
Comment (josephschorr) [inline .github/workflows/lint.yaml:7]: Why was this changed?

Comment (vroldanbet) [reply .github/workflows/lint.yaml:7]: So we don't get duplicated CI builds running in a PR. It's the same pattern we use in other repos.

Comment (josephschorr) [reply .github/workflows/lint.yaml:7]: Can you put that as a distinct commit then and explain that in the message?

Comment (vroldanbet) [reply .github/workflows/lint.yaml:7]: addressed in https://github.com/authzed/api/pull/40/commits/1a3254e5062a5d65b09a8ef048ca38d6f2464adb

**Review by josephschorr — APPROVED** [review]
LGTM

File: authzed/api/v1/permission_service.proto [hunk 6]
```diff
@@ -250,13 +251,25 @@ message CheckPermissionRequest {
 
   // subject is the subject that will be checked for the permission or relation.
   SubjectReference subject = 4 [ (validate.rules).message.required = true ];
+
+  /** context consists of named values that are injected into the caveat evaluation context **/
+  google.protobuf.Struct context = 5 [ (validate.rules).message.required = false ];
```
[the remark this answers is not in the document]
Comment (josephschorr) [reply authzed/api/v1/permission_service.proto:256]: You can't have a raw map in Protocol Buffers. It would have to be that all keys are string typed or somesuch.

[the remark this answers is not in the document]
Comment (jzelinskie) [reply authzed/api/v1/permission_service.proto:256]: But the struct is just a message that contains a map, why not dump the struct and just use the map?

[the remark this answers is not in the document]
Comment (josephschorr) [reply authzed/api/v1/permission_service.proto:256]: https://pkg.go.dev/google.golang.org/protobuf/types/known/structpb provides a lot of nicer helper code around handling of the type, as it is a "standard", more or less

[the remark this answers is not in the document]
Comment (jzelinskie) [reply authzed/api/v1/permission_service.proto:256]: neat, i hope other languages have this, too

[the remark this answers is not in the document]
Comment (jzelinskie) [reply authzed/api/v1/permission_service.proto:256]: is this the first google type in the user facing path?

[the remark this answers is not in the document]
Comment (vroldanbet) [reply authzed/api/v1/permission_service.proto:256]: @jzelinskie just inspected API and yes, beyond `google/api/annotations.proto`, it seems to be the first time we use a "Well Known Type" in the API

[the remark this answers is not in the document]
Comment (jzelinskie) [reply authzed/api/v1/permission_service.proto:256]: Is that going to be annoying for folks to construct in each language? I'd hate to tell folks they need even more libraries just to perform a check request.

[the remark this answers is not in the document]
Comment (vroldanbet) [reply authzed/api/v1/permission_service.proto:256]: honestly I don't know, we'd need to check how it's done in each language. What would be acceptable from your perspective? Would creating our own type be any less annoying?

File: .github/workflows/lint.yaml (hunk 5 above)
Comment (jzelinskie) [reply .github/workflows/lint.yaml:7]: Ideally, development would be done on a fork and this value should be `*` so that if we were to produce a v2 branch etc... it would run CI correctly.

**Review by jzelinskie — APPROVED** [review]

**Review by josephschorr — APPROVED** [review]
